src/tasks.ts

```diff
@@ -96,59 +114,120 @@ export async function main(): Promise<void> {
     const tree = buildTree(leafs);
     debug('Report tree root', tree.root);
 
-    debug('Simulating the report');
-    await Shared.ORACLE.connect(Shared.SIGNER).submitReport.staticCall(
-        targetEpoch,
-        tree.root,
+    debug('Saving the artifact');
+    const artifact: Artifact = {
         distributed,
-        '', // CID TBA, but can be precalculated
+        sourceEpoch,
+        targetEpoch,
+        tree,
+    };
+    const filename = saveArtifact(artifact);
+    debug(`Artifact saved to ${filename}`);
+
+    // Uploading the tree optimistically
+    debug('Uploading the artifact to IPFS');
+    const cid = await uploadFile(filename);
+    debug('Tree uploaded, CID is', cid.IpfsHash);
+
+    const report = {
+        consensusVersion: Shared.CONSENSUS_VERSION,
+        refSlot: refSlot,
+        treeRoot: tree.root,
+        treeCid: cid.IpfsHash,
+        distributed: distributed,
+    };
+    const reportHash = hashReport(report);
+
+    const { currentFrameMemberReport } =
+        await Shared.HASHCONSENSUS.getConsensusStateForMember(
+            Shared.SIGNER.address,
+        );
+    if (currentFrameMemberReport == reportHash) {
+        debug('Provided hash already submitted');
+        return;
+    }
+
+    debug('Simulating sending report hash');
+    await Shared.HASHCONSENSUS.connect(Shared.SIGNER).submitReport.staticCall(
+        refSlot,
+        reportHash,
+        Shared.CONSENSUS_VERSION,
     );
     debug('Simulation successfull!');
 
-    debug('Uploading the tree to IPFS');
-    const cid = await uploadTree(tree);
-    debug('Tree uploaded, CID is', cid.toString());
-
     debug('Sending the report');
-    await Shared.ORACLE.connect(Shared.SIGNER).submitReport(
-        targetEpoch,
-        tree.root,
-        distributed,
-        cid.toString(),
+    await Shared.HASHCONSENSUS.connect(Shared.SIGNER).submitReport(
+        refSlot,
+        reportHash,
+        Shared.CONSENSUS_VERSION,
     );
-
     debug('Report sent, storing the artifact');
-    saveArtifact(from, to, tree);
+
+    const { currentFrameConsensusReport } =
+        await Shared.HASHCONSENSUS.getConsensusStateForMember(
+            Shared.SIGNER.address,
+        );
+
+    if (currentFrameConsensusReport == ZeroHash) {
+        debug('No consensus reached');
+        return;
+    }
+
+    if (currentFrameConsensusReport != reportHash) {
+        debug('Oracle`s hash differs from consensus report hash. Exiting');
+        return;
+    }
```

Comment:
I guess there should be sophisticated algorithm in the future just because of the fact that reports from different members may not be sent as quickly or they may not be included in the blocks by that time

Reply:
Currently, it's just the first member which reached consensus will send a report, because no fast-lane implemented. With that in place it will require some changes, indeed.